File: src/main/java/com/mbayou/kick4k/api/ApiClient.java
package com.mbayou.kick4k.api;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.mbayou.kick4k.KickConfiguration;
import com.mbayou.kick4k.authorization.AuthorizationClient;

import java.io.IOException;
import java.lang.reflect.Array;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.StringJoiner;

public abstract class ApiClient {
    protected final HttpClient httpClient;
    protected final ObjectMapper mapper;
    protected final KickConfiguration configuration;
    protected final AuthorizationClient authorization;

    protected ApiClient(HttpClient httpClient, ObjectMapper mapper, KickConfiguration configuration, AuthorizationClient authorization) {
        this.httpClient = httpClient;
        this.mapper = mapper;
        this.configuration = configuration;
        this.authorization = authorization;
    }

    protected RequestBuilder get(String path) {
        return new RequestBuilder("GET", path);
    }

    protected RequestBuilder post(String path) {
        return new RequestBuilder("POST", path);
    }

    protected RequestBuilder patch(String path) {
        return new RequestBuilder("PATCH", path);
    }

    protected RequestBuilder delete(String path) {
        return new RequestBuilder("DELETE", path);
    }

    public class RequestBuilder {
        private final String method;
        private String path;
        private Map<String, Object> queryParams;
        private Object bodyObject;
        private Class<?> bodyClass;

        public RequestBuilder(String method, String path) {
            this.method = method;
            this.path = path;
        }

        public RequestBuilder queryParams(Map<String, Object> queryParams) {
            this.queryParams = queryParams;
            return this;
        }

        public RequestBuilder queryParams(Object queryParams) {
            if (queryParams == null) {
                return this;
            }

            Map<String, Object> map = mapper.convertValue(queryParams, new TypeReference<>() {});
            map.entrySet().removeIf(e -> e.getValue() == null);

            if (this.queryParams == null) {
                this.queryParams = map;
            } else {
                this.queryParams.putAll(map);
            }

            return this;
        }

        public RequestBuilder pathParams(Map<String, Object> params) {
            for (Map.Entry<String, Object> entry : params.entrySet()) {
                String placeholder = "\\{" + entry.getKey() + "\\}";
                String encodedValue = encode(entry.getValue().toString());
                this.path = this.path.replaceAll(placeholder, encodedValue);
            }
            return this;
        }

        public RequestBuilder body(Object bodyObject) {
            this.bodyObject = bodyObject;
            return this;
        }

        public <T> T send(TypeReference<ApiResponse<T>> typeRef) {
            try {
                String url = buildUrl(configuration.getBaseUrl() + this.path, this.queryParams);
                HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
                        .uri(URI.create(url))
                        .header("Authorization", "Bearer " + authorization.getAccessToken())
                        .header("Accept", "application/json");

                if ("GET".equalsIgnoreCase(method)) {
                    requestBuilder.GET();
                } else {
                    String jsonBody = this.bodyObject == null ? "" : mapper.writeValueAsString(this.bodyObject);
                    requestBuilder.header("Content-Type", "application/json");
                    requestBuilder.method(this.method, HttpRequest.BodyPublishers.ofString(jsonBody));
                }

                HttpResponse<String> response = httpClient.send(requestBuilder.build(), HttpResponse.BodyHandlers.ofString());
                String body = response.body();
                if (body == null || body.isEmpty()) {
                    return null;
                }

                ApiResponse<T> apiResponse = mapper.readValue(response.body(), typeRef);
                if (!apiResponse.isSuccess() && response.statusCode() >= 400) {
                    throw new ApiException(response.statusCode(), apiResponse.getMessage());
                }

                return apiResponse.getData();
            } catch (IOException | InterruptedException exception) {
                Thread.currentThread().interrupt();
                throw new RuntimeException("Failed to send API request", exception);
            }
        }
    }

    private static String buildUrl(String path, Map<String, Object> queryParams) {
        if (queryParams == null) {
            return path;
        }

        StringJoiner joiner = new StringJoiner("&", path + "?", "");
        for (Map.Entry<String, Object> param : queryParams.entrySet()) {
            Object value = param.getValue();

            if (value instanceof Iterable<?>) {
                for (Object v : (Iterable<?>) value) {
                    joiner.add(encode(param.getKey()) + "=" + encode(String.valueOf(v)));
                }
            } else if (value != null && value.getClass().isArray()) {
                int length = Array.getLength(value);

                for (int i = 0; i < length; i++) {
                    Object v = Array.get(value, i);
                    joiner.add(encode(param.getKey()) + "=" + encode(String.valueOf(v)));
                }
            } else {
                joiner.add(encode(param.getKey()) + "=" + encode(String.valueOf(value)));
            }
        }

        return joiner.toString();
    }

    private static String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }
}

File: src/main/kotlin/com/mbayou/kick4k/api/ApiClient.kt
package com.mbayou.kick4k.api

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import com.mbayou.kick4k.KickConfiguration
import com.mbayou.kick4k.authorization.AuthorizationClient
import java.io.IOException
import java.lang.reflect.Array
import java.net.URI
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.charset.StandardCharsets
import java.util.StringJoiner

abstract class ApiClient(
    protected val httpClient: HttpClient,
    protected val mapper: ObjectMapper,
    protected val configuration: KickConfiguration,
    protected val authorization: AuthorizationClient,
) {
    protected fun get(path: String): RequestBuilder = RequestBuilder("GET", path)
    protected fun post(path: String): RequestBuilder = RequestBuilder("POST", path)
    protected fun patch(path: String): RequestBuilder = RequestBuilder("PATCH", path)
    protected fun delete(path: String): RequestBuilder = RequestBuilder("DELETE", path)

    inner class RequestBuilder(private val method: String, private var path: String) {
        private var queryParams: MutableMap<String, Any?>? = null
        private var bodyObject: Any? = null

        fun queryParams(params: Map<String, Any?>?): RequestBuilder = apply {
            if (params != null) {
                if (queryParams == null) {
                    queryParams = params.toMutableMap()
                } else {
                    queryParams!!.putAll(params)
                }
            }
        }

        fun queryParams(params: Any?): RequestBuilder = apply {
            if (params == null) return@apply
            val map: MutableMap<String, Any?> =
                mapper.convertValue(params, object : TypeReference<MutableMap<String, Any?>>() {})
            map.entries.removeIf { it.value == null }
            if (queryParams == null) {
                queryParams = map
            } else {
                queryParams!!.putAll(map)
            }
        }

        fun pathParams(params: Map<String, Any?>): RequestBuilder = apply {
            params.forEach { (key, value) ->
                val placeholder = "\\{" + key + "\\}"
                val encodedValue = encode(value?.toString() ?: "")
                path = path.replace(placeholder.toRegex(), encodedValue)
            }
        }

        fun body(bodyObject: Any?): RequestBuilder = apply { this.bodyObject = bodyObject }

        fun <T> send(typeRef: TypeReference<ApiResponse<T>>): T? {
            val url = buildUrl(configuration.baseUrl + path, queryParams)
            val requestBuilder = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer ${authorization.getAccessToken()}")
                .header("Accept", "application/json")

            if ("GET".equals(method, ignoreCase = true)) {
                requestBuilder.GET()
            } else {
                val jsonBody = bodyObject?.let { mapper.writeValueAsString(it) } ?: ""
                requestBuilder.header("Content-Type", "application/json")
                    .method(method, HttpRequest.BodyPublishers.ofString(jsonBody))
            }

            val response = sendRequest(requestBuilder)
            val body = response.body()
            if (body.isNullOrEmpty()) {
                return null
            }

            val apiResponse = mapper.readValue(body, typeRef)
            if (!apiResponse.isSuccess() && response.statusCode() >= 400) {
                throw ApiException(response.statusCode(), apiResponse.message)
            }
            return apiResponse.data
        }

        private fun sendRequest(requestBuilder: HttpRequest.Builder): HttpResponse<String> {
            return try {
                httpClient.send(requestBuilder.build(), HttpResponse.BodyHandlers.ofString())
            } catch (exception: InterruptedException) {
                Thread.currentThread().interrupt()
                throw RuntimeException("Failed to send API request", exception)
            } catch (exception: IOException) {
                throw RuntimeException("Failed to send API request", exception)
            }
        }
    }

    companion object {
        private fun buildUrl(path: String, queryParams: Map<String, Any?>?): String {
            if (queryParams.isNullOrEmpty()) {
                return path
            }

            val joiner = StringJoiner("&", "$path?", "")
            queryParams.forEach { (key, value) ->
                when {
                    value == null -> Unit
                    value is Iterable<*> -> value.forEach { joiner.add("${encode(key)}=${encode(it.toString())}") }
                    value.javaClass.isArray -> {
                        val length = Array.getLength(value)
                        for (i in 0 until length) {
                            val element = Array.get(value, i)
                            joiner.add("${encode(key)}=${encode(element.toString())}")
                        }
                    }
                    else -> joiner.add("${encode(key)}=${encode(value.toString())}")
                }
            }
            return joiner.toString()
        }

        private fun encode(value: String): String = URLEncoder.encode(value, StandardCharsets.UTF_8)
    }

    protected inline fun <reified T> responseType(): TypeReference<ApiResponse<T>> =
        object : TypeReference<ApiResponse<T>>() {}
}
